### PR TITLE
Move thread pool initialization to native side

### PR DESCRIPTION
**Pull request by kouvel**
- Following up from https://github.com/dotnet/runtime/pull/36697. Currently the initialization occurs when a static variable is accessed before calling into the VM, and after thinking about it some more the intent and need to initialize the thread pool can be a bit unclear
- The initialization is specific to this implementation and the managed side doesn't necessarily need to know about it
- Moved the initialization to the native side similarly to other thread pool APIs
- Added some more assertions to relevant managed-to-native entry point paths to ensure that they are not called before the thread pool is initialized
- This adds a non-volatile check when requesting a worker thread, which is already a very slow path. I ran into some issues with testing on the arm64 machine with updated locally built libcoreclr.so, so wasn't able to test that. I looked at the baseline profile and looking at time spent exclusively in ThreadPoolNative::RequestWorkerThread() and what it already does, I don't think the change would result in any significant perf difference.

### CODE REVIEW

**Comment by kouvel** [comment]
CC @benaadams @janvorli @stephentoub

File: src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.Portable.cs [hunk 1]
```diff
@@ -317,6 +317,8 @@ internal void WaitForRemoval()
 
     public static partial class ThreadPool
     {
+        internal const bool s_enableWorkerTracking = false;
```
Comment (benaadams) [inline src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.Portable.cs:320]:
```suggestion
        internal static readonly bool s_enableWorkerTracking = false;
```
As a `const` C# gives unreachable code warnings

**Review by benaadams — APPROVED** [review]
One compile error nit

File: src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.Portable.cs (hunk 1 above)
Comment (kouvel) [reply src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.Portable.cs:320]: Ah ok, thanks!

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.Portable.cs]: We'll get a warning either way: if it's const, you might get compiler warnings about unreachable code, and if it's a static readonly, you'll get analyzer warnings about a static readonly being initialized to a constant value.  Another alternative is to make it a property:
```C#
internal static bool EnableWorkerTracking { get; } = GetEnableWorkerTracking();
```
on coreclr and:
```C#
internal static bool EnableWorkerTracking => false;
```
on mono.  But you could also just make it a const and suppress any compiler warnings.

Comment (kouvel) [reply src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.Portable.cs]: Thanks, I have suppressed the warning

**Review by akoeplinger — APPROVED** [review]

**Review by stephentoub — APPROVED** [review]

**Pull request merged**